packages/api/src/controllers/taskController.js

```diff
@@ -58,6 +62,52 @@ async function getTaskAssigneeAndFinalWage(farm_id, user_id, task_id) {
   return { assignee_user_id, finalWage };
 }
 
+async function formatAnimalMovementTaskForDB(data) {
+  if (data.locations && data.locations.length > 1) {
+    throw customError('Only one location can be assigned to this task type', 400);
+  }
+
+  if (!data.animal_movement_task) {
+    return data;
+  }
+
+  if (data.animal_movement_task.purpose_relationships) {
+    throw customError(
+      `Invalid field: "purpose_relationships" should not be included. Use "purposes" instead`,
+    );
+  }
+
+  if (!data.animal_movement_task.purposes) {
+    delete data.animal_movement_task.purposes;
+    return data;
+  }
+
+  checkIsArray(data.animal_movement_task.purposes, 'purposes');
+
+  const formattedPurposes = [];
+
+  if (data.animal_movement_task.purposes.length) {
+    const purposes = await AnimalMovementPurposeModel.query();
+    const purposesMap = purposes.reduce((map, { key, id }) => ({ ...map, [key]: id }), {});
+    for (const { key, other_purpose } of data.animal_movement_task.purposes) {
+      if (!purposesMap[key]) {
+        throw customError(`Purpose key "${key}" is not supported`);
+      }
+      const formattedPurpose = {
+        purpose_id: purposesMap[key],
+        ...(other_purpose ? { other_purpose } : {}),
+      };
+
+      formattedPurposes.push(formattedPurpose);
+    }
+  }
+
+  data.animal_movement_task.purpose_relationships = formattedPurposes;
```

Comment:
Is there a reason we can't have the FE send out the purpose IDs instead? I was thinking of something similar to `related_animal_ids`, where we would have `purpose_ids = [1, 2, 3]` and then the BE wouldn't need to do a search by key, since the FE already has both IDs and keys

Reply:
I don't entirely remember the conversation exactly so sorry if I said something different, but here's what I'm thinking:
- We fetch the list of purposes on the FE from the API, containing both IDs and keys, so we get something like `purposes = [{ id: 1, key: 'GRAZING' }, { id: 5, key: 'OTHER' }]`
- On the FE, we check for the "OTHER" key in case where we have some logic to display certain fields if that option was selected.
- Once we're ready to send the info back to the API and save the task, we send only the purposes IDs, and the "other purpose" note, if it has been shown to the user per the logic above and filled out.
- The BE takes the list of purpose IDs and checks if the key for any of them is "OTHER". If it is, it allows the "other purpose" field and stores it in the DB. If none of the purposes matches that key, it either returns a bad request to inform the client the request doesn't make sense, or simply ignores the other purpose string and doesn't store it. The DB constraint will disallow the BE from storing something that doesn't make sense.
This way we're preventing inconsistencies on all 3 layers:
- On the FE, by only showing the field when it's relevant.
- On the BE, by not passing through data to the DB that doesn't make sense.
- On the database, by disallowing storing data that doesn't make sense.
The FE however doesn't need to worry about checking what the ID is for "Other", it just deals with keys on the visual side and sends IDs once it sends the info for the BE to take over.
Let me know if this makes sense or if I'm missing something!

Reply:
You may have meant not to send relationships from the beginning. Sending IDs and other_purpose separately works well, so I'll make the necessary adjustments!